base_layer/p2p/src/services/domain_deserializer.rs

```diff
@@ -0,0 +1,111 @@
+// Copyright 2019 The Tari Project
+//
+// Redistribution and use in source and binary forms, with or without modification, are permitted provided that the
+// following conditions are met:
+//
+// 1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following
+// disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
+// following disclaimer in the documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote
+// products derived from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+// INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+// USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use futures::{Future, Poll};
+use serde::export::PhantomData;
+use tari_comms::{
+    domain_subscriber::MessageInfo,
+    message::{DomainMessageContext, MessageError},
+};
+use tari_utilities::message_format::MessageFormat;
+use tokio_threadpool::{blocking, BlockingError};
+
+/// Future which asynchonously attempts to deserialize DomainMessageContext into
+/// a `(MessageInfo, T)` tuple where T is [MessageFormat].
+pub struct DomainMessageDeserializer<T> {
+    message: Option<DomainMessageContext>,
+    _t: PhantomData<T>,
+}
+
+impl<T> DomainMessageDeserializer<T> {
+    /// Create a new DomainMessageDeserializer from the given DomainMessageContext
+    pub fn new(message: DomainMessageContext) -> Self {
+        Self {
+            message: Some(message),
+            _t: PhantomData,
+        }
+    }
+}
+
+impl<T: MessageFormat> Future for DomainMessageDeserializer<T> {
+    type Error = BlockingError;
+    type Item = Result<(MessageInfo, T), MessageError>;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        let msg = self.message.take().expect("poll called twice on Deserializer");
+        blocking(|| {
```

Comment:
This seems overkill to need blocking here? It's just a memory mapping? This is definitely a discussion to have to decide where the threshold is for an operation to be put into Blocking but a straight up memory based read seems like its plenty fast enough to be done directly?

Reply:
Yup - maybe - not sure either - it's a memory mapping because we happen to be using bincode - however that could change and is up to the MessageFormat impl. Tasks are super cheap to create according to tokio. Definitely up for debate and I'm sure well come across many of these cases.